### PR TITLE
fix: handle null case in ResizeObserver callback

### PR DESCRIPTION
**Pull request by spmonahan**
## Previous Behavior

Partner telemetry is getting an error, `TypeError: e[0] is undefined` in `useMessageBarReflow.js`.

There are two instances of accessing the zeroth element of an array in this file:

1. [entries](https://github.com/microsoft/fluentui/blob/master/packages/react-components/react-message-bar/library/src/components/MessageBar/useMessageBarReflow.ts#L29)
2. [borderBoxSize](https://github.com/microsoft/fluentui/blob/master/packages/react-components/react-message-bar/library/src/components/MessageBar/useMessageBarReflow.ts#L31)

According to MDN's example, `borderBoxSize` may be empty: https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry/borderBoxSize#examples

It's not clear from [MDN's docs on ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver/ResizeObserver#callback) that this is also the case for the callback's `entries` parameter but we handle the empty case by a nullish check on the `entry` already so this is unlikely to be the problem.

## New Behavior

Adds a check to ensure there is an element at index `0` in `borderBoxSize`.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports | Baseline (minified/GZIP) | PR    | Change     |
| :---------------- | -----------------------: | ----: | ---------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ReactComponents.fixture.js'>react-components: entire library</abbr> | `1.1 MB`<br />`272.081 kB` | `1.1 MB`<br />`272.069 kB` | `-42 B` <br />`-12 B` |
| <samp>react-message-bar</samp> <br /> <abbr title='bundle-size/Default.fixture.js'>MessageBar (all components)</abbr> | `24.85 kB`<br />`9.256 kB` | `24.808 kB`<br />`9.245 kB` | `-42 B` <br />`-11 B` |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports | Size (minified/GZIP) |
| ----------------- | -------------------: |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr> | `69.14 kB`<br />`20.137 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `218.317 kB`<br />`63.258 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr> | `44.447 kB`<br />`14.59 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr> | `8.39 kB`<br />`2.64 kB` |
| <samp>react-timepicker-compat</samp> <br /> <abbr title='bundle-size/TimePicker.fixture.js'>TimePicker</abbr> | `107.387 kB`<br />`35.758 kB` |
</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/5bea3fa901c841e41438fff4b7ae6b38bf30b982'>5bea3fa901c841e41438fff4b7ae6b38bf30b982</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/Avatar-mount_base.html">611</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/Avatar-mount.html">619</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/Button-mount_base.html">295</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/Button-mount.html">302</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/Field-mount_base.html">1143</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/Field-mount.html">1146</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/FluentProvider-mount_base.html">719</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/FluentProvider-mount.html">694</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/FluentProviderWithTheme-mount_base.html">80</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/FluentProviderWithTheme-mount.html">88</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">34</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">42</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">80</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">80</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/MakeStyles-mount_base.html">857</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/MakeStyles-mount.html">886</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/Persona-mount_base.html">1752</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/Persona-mount.html">1732</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/SpinButton-mount_base.html">1434</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/SpinButton-mount.html">1406</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SwatchPicker</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/SwatchPicker-mount_base.html">1647</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/32971/perf-test-react-components/SwatchPicker-mount.html">1720</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

File: change/@fluentui-react-message-bar-284ced69-f89c-4c09-814b-63c9a7cff1bb.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (fabricteam) [inline change/@fluentui-react-message-bar-284ced69-f89c-4c09-814b-63c9a7cff1bb.json:1]: ## 🕵🏾‍♀️ [<b data-uid='vrt-gist'></b> visual regressions to review in the fluentuiv9 Visual Regression Report](https://staticpagerelease.z16.web.core.windows.net/azurestaticapps/vr-approval/index.html?organization=uifabric&project=fabricpublic&prId=32971&commitId=44a8fd71818ba6b5b4ff91f235df4a9f423fefa8&env=prod&clientType=FLUENTUI)


  <details><summary><b>Avatar Converged</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Avatar Converged.badgeMask.chromium.png|4| Changed|
|Avatar Converged.Badge Mask RTL.chromium.png|1| Changed|

</details>


  <details><summary><b>Drawer</b> 2 screenshots</summary>

|Image Name|Diff(in Pixels)| Image Type
|-|-|-|
|Drawer.Full Overlay Dark Mode.chromium.png|984| Changed|
|Drawer.overlay drawer full.chromium.png|1147| Changed|

</details>




<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-message-bar/library/src/components/MessageBar/useMessageBarReflow.ts [hunk 2]
```diff
@@ -27,9 +27,20 @@ export function useMessageBarReflow(enabled: boolean = false) {
...
+
+      if (!entry) {
+        return;
+      }
+
       // `borderBoxSize` is not supported before Chrome 84, Firefox 92, nor Safari 15.4
-      const inlineSize = entry?.borderBoxSize?.[0].inlineSize ?? entry?.target.getBoundingClientRect().width;
-      if (inlineSize === undefined || !entry) {
+      let inlineSize: number | undefined;
+      if (entry.borderBoxSize && entry.borderBoxSize.length > 0) {
```
Comment (MLoughry) [inline packages/react-components/react-message-bar/library/src/components/MessageBar/useMessageBarReflow.ts:37]: Nit for conciseness
```suggestion
      if (entry.borderBoxSize?.length) {
```

Comment (MLoughry) [reply packages/react-components/react-message-bar/library/src/components/MessageBar/useMessageBarReflow.ts:37]: Actually, I'm not sure what this if-else block is meant to do not already done by the code replacing it. They're functionally equivalent.

Comment (spmonahan) [reply packages/react-components/react-message-bar/library/src/components/MessageBar/useMessageBarReflow.ts:37]: The difference is that the previous code nullish checked `borderBoxSize` and the new code does that and also checks that `borderBoxSize` has elements.

Until this issue came up I was unaware `borderBoxSize` [might be empty](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry/borderBoxSize#examples) in which case accessing `borderBoxSize?.[0].inlineSize` would throw the observed error.

I like your conciseness change!

**Review by layershifter — APPROVED** [review]

File: packages/react-components/react-message-bar/library/src/components/MessageBar/useMessageBarReflow.ts (hunk 2 above)
Comment (MLoughry) [reply packages/react-components/react-message-bar/library/src/components/MessageBar/useMessageBarReflow.ts]: @spmonahan No, if an optional chaining operator `?.` is used on a property or array reference (as in `borderBoxSize?.[0]`), then if the object of the lookup is `null` or `undefined`, then the rest of the lookup is skipped.

```ts
if (borderBoxSize === undefined) {
  assert(borderBoxSize?.[0].inlineSize === undefined); // No Type error
}
```

You can try this yourself with `node -i` or the browser dev tools.

Comment (spmonahan) [reply packages/react-components/react-message-bar/library/src/components/MessageBar/useMessageBarReflow.ts]: @MLoughry The issue arises when `borderBoxSize` is defined but an empty array:

```js
let borderBoxSize = [];
assert(borderBoxSize?.[0].inlineSize === undefined); // <-- Type error
```

To handle the empty array case we either need my existing change or:

```js
let borderBoxSize = [];
assert(borderBoxSize?.[0]?.inlineSize === undefined); // No Type error
```

msedge_pVH6NHsxa3


```

**Pull request merged**